File: frontend/src/utils/rut.ts
const CLEAN_RE = /[\.\s]/g

export function cleanRut(rut: string): string {
  return rut.replace(CLEAN_RE, '').toUpperCase()
}

export function validateRut(rut: string): boolean {
  const cleaned = cleanRut(rut)
  if (!/^\d{1,8}-[\dK]$/.test(cleaned)) return false
  const [digits, dv] = [cleaned.slice(0, -2), cleaned.slice(-1)]
  let total = 0
  let factor = 2
  for (let i = digits.length - 1; i >= 0; i--) {
    total += parseInt(digits[i]) * factor
    factor = factor < 7 ? factor + 1 : 2
  }
  const remainder = total % 11
  const expected = 11 - remainder
  const expectedDv = expected === 11 ? '0' : expected === 10 ? 'K' : String(expected)
  return dv === expectedDv
}

export function formatRut(rut: string): string {
  const cleaned = cleanRut(rut)
  const match = cleaned.match(/^(\d+)-?([\dK])$/)
  if (!match) return rut
  const [, digits, dv] = match
  const formatted = digits.replace(/\B(?=(\d{3})+(?!\d))/g, '.')
  return `${formatted}-${dv}`
}

File: backend/app/utils/rut.py
import re

_CLEAN_RE = re.compile(r'[\.\s]')


def clean_rut(rut: str) -> str:
    """Strip dots/spaces, uppercase, e.g. '76.543.210-k' -> '76543210-K'"""
    return _CLEAN_RE.sub('', rut).upper()


def validate_rut(rut: str) -> bool:
    """Return True if Chilean RUT format and check digit are valid."""
    cleaned = clean_rut(rut)
    if not re.fullmatch(r'\d{1,8}-[\dK]', cleaned):
        return False
    digits, dv = cleaned[:-2], cleaned[-1]
    total = 0
    factor = 2
    for c in reversed(digits):
        total += int(c) * factor
        factor = factor + 1 if factor < 7 else 2
    remainder = total % 11
    expected = 11 - remainder
    if expected == 11:
        expected_dv = '0'
    elif expected == 10:
        expected_dv = 'K'
    else:
        expected_dv = str(expected)
    return dv == expected_dv
